### ISSUE TITLE
GPO Wish: Copy cert_override.txt from share to current profile

### ISSUE THREAD

**Issue by cypherstream** [mozilla/policy-templates#360]
In a corporate environment we have some applications that use self signed certificates.  After Firefox 53 even if these self signed certificates are deployed via GPO (with Firefox set to use the Windows Certificate libraries), they still show as untrusted.  After Firefox 53 it seems Mozilla decided to mark ALL self signed certificates as untrusted even if they were purposesly loaded into Firefox itself, or the Windows trusted root certificate authority.  Only Internet Explorer will trust these certs.

For this reason to aid in finally moving 100% away from Internet Explorer, we need a way to get a custom cert_override.txt file to every single user in the corporate environment.  How this works is an IT VM or loaner machine would go out to the specific sites required, all adding the exceptions.  Then they would grab their local cert_override.txt out of their firefox profile at %AppData%\Roaming\Mozilla\Firefox\Profiles\xxxxxxxx.default\ and copy it to a publically accessible share.  Finally in the GPO they would set this to enabled and paste in the UNC path along with the file name.  

Because every user on every machine has a very unique firefox profile directory name, it has been seen to be impossible to simply write a script to copy a cert_override.txt from one user to another, because it is impossible to get the path right since it is unique for every user and computer.

A GPO to do this for us would save a lot of headache and help us move more sites away from Internet Explorer, which is now known as a "compatibility solution" not a browser.

**Comment by mkaply** [comment]
So FYI, the self signed problem is because we now require self signed certificates to be chained to a root. We're investigating undoing that change.

**Comment by cypherstream** [comment]
Mike,

I'm using your script for now to copy a cert_override.txt to the Firefox local profile as found on your site here:
https://mike.kaply.com/2016/05/24/default-profile-directory-doesnt-work-in-firefox-46/

Many thanks to your time and efforts working on this platform!

Basically what I did was replace this certDBFile.append("cert8.db") with this certDBFile.append("cert_override.txt").

Group Policy Preferences copies a centralized cert_override.txt down to C:\Program Files\Mozilla Firefox\defaults\profile.

Your script, which has been added as our second .cfg file in our autoconfig.js that we push down with GPP now sees if cert_override.txt does not exist in the Firefox profile, it copies it from that defaults location.  This seems to work for the majority of what we need.  The only caveat is that there isn't a method to update the cert_override.txt.  Once it already exists in the profile, it will not overwrite.  Someone who may have already added an exception for a site will not get the companies list (which currently is 4 sites).  

We tried other means to script but with the unique naming of the Firefox profile, I couldn't figure out how to get it to work.

Even if Mozilla made the self signed certificates chained to root a preference that could be toggled on or off, I could live with that as well.  Simply because we imported these self signed certificates as trusted root certs using Group Policy and obviously we have Firefox ESR configured to utilize the Windows certificate store.  Our first .cfg file specifies many pref's and it would not be difficult to add one more.  Every group policy refresh copies the latest version of this file down to where it needs to go on the local computer from our corporate file share.

Thanks for your efforts!

**Comment by mkaply** [comment]
So FYI, we've fixed the self signed chaining problem.

Could you try a Firefox nightly build and see if things are working better when you import certs from the OS?

**Comment by cypherstream** [comment]
> Could you try a Firefox nightly build and see if things are working better when you import certs from the OS?

Yes I finally had a moment to grab the nightly build and install it.  It appears that the new build trusts self signed certs that are purposely imported into the Trusted Root certificates via GPO's.  When this makes it to ESR it will ease a lot of corporate requirements where specific software is secured, but via a self signed certificate.  In most cases using group policy to deploy these certificates is the answer and this change allows us to finally move away from Internet Explorer 11 for these types of applications and onto something MUCH better... Firefox.

I always appreciate your contribution to the project.  Thank you for taking feedback and working with end users who have to manage and maintain Firefox on a large scale.

**Comment by cypherstream** [comment]
> Could you try a Firefox nightly build and see if things are working better when you import certs from the OS?

Yes I finally had a moment to grab the nightly build and install it.  It appears that the new build trusts self signed certs that are purposely imported into the Trusted Root certificates via GPO's.  When this makes it to ESR it will ease a lot of corporate requirements where specific software is secured, but via a self signed certificate.  In most cases using group policy to deploy these certificates is the answer and this change allows us to finally move away from Internet Explorer 11 for these types of applications and onto something MUCH better... Firefox.

I always appreciate your contribution to the project.  Thank you for taking feedback and working with end users who have to manage and maintain Firefox on a large scale.

**Comment by cypherstream** [comment]
> Could you try a Firefox nightly build and see if things are working better when you import certs from the OS?

Yes I finally had a moment to grab the nightly build and install it.  It appears that the new build trusts self signed certs that are purposely imported into the Trusted Root certificates via GPO's.  When this makes it to ESR it will ease a lot of corporate requirements where specific software is secured, but via a self signed certificate.  In most cases using group policy to deploy these certificates is the answer and this change allows us to finally move away from Internet Explorer 11 for these types of applications and onto something MUCH better... Firefox.

I always appreciate your contribution to the project.  Thank you for taking feedback and working with end users who have to manage and maintain Firefox on a large scale.
